torch_xla/csrc/helpers.cpp

```diff
@@ -254,27 +254,23 @@ xla::XlaOp XlaHelpers::ReshapeToRank(xla::XlaOp input, int64_t expected_rank,
 absl::optional<XlaHelpers::DynamicReshapeInfo>
 XlaHelpers::GetDynamicReshapeInfo(const xla::Shape& input_shape,
                                   absl::Span<const int64_t> output_sizes) {
-  int64_t input_dynamic_dimension = GetDynamicDimension(input_shape);
-  if (input_dynamic_dimension < 0) {
+  int64_t input_dyndim_idx = GetDynamicDimension(input_shape);
+  if (input_dyndim_idx < 0) {
     return absl::nullopt;
   }
   DynamicReshapeInfo info;
   info.output_shape =
       xla::ShapeUtil::MakeShape(input_shape.element_type(), output_sizes);
   if (info.output_shape.rank() > 0) {
-    int64_t size_at_dyndim = 1;
-    for (int64_t i = 0; i <= input_dynamic_dimension; ++i) {
-      size_at_dyndim *= input_shape.dimensions(i);
+    int64_t size_prod_until_dyndim = 1;
+    for (int64_t i = 0; i <= input_dyndim_idx; ++i) {
+      size_prod_until_dyndim *= input_shape.dimensions(i);
     }
     int64_t dynamic_dimension = -1;
     int64_t out_size = 1;
     for (int64_t i = 0; i < output_sizes.size(); ++i) {
-      XLA_CHECK_LE(out_size, size_at_dyndim / input_shape.dimensions(
```

Comment:
why is this check removed?

Reply:
It's the trickest part of the PR. Let's use the test `test_view_copy_symint_negative_input_shape` for example. In the test, we do `t3=t2.view(2,3)` where t2.shape=torch.Size([<=6, 1]) with real size [6, 1]. To my understanding, the point of this function `XlaHelpers::GetDynamicReshapeInfo` is to find out which output tensor dimension is dynamic. Theoretically, we should be able to reshape a [<=6,1] tensor to a [2, <=3] tensor but in reality this `XLA_CHECK_LE` would fail for no good reason: when i=1, out_size=2, size_at_dyndim/input_shape.dimensions(input_dynamic_dimension)=6/6=1

Addditionally, after I comment out this check, the CI is green so it seems to me we're not gaining much from the check. And the original PR https://github.com/pytorch/xla/pull/1496 doesn't have tests for this check..

Reply:
Now that the `view` op doesn't depends on this function, I'd still think we can remove this check, for the above reason.

Reply:
is this true through? I don't think we can always reshape [<=6,1] tensor to a [2, <=3], what if it is a `[3,1]`, there is no way for us to reshape it [2, x].

Reshaping dynamic tensor is tricky, I wonder if XLA has any guideline on this.

Reply:
ok, I added the check back, since this PR doesn't necessarily need to use this function. I can deal with it later if it becomes a blocker.

Though I'd like to keep the renaming change as it's really confusing to read this part of code.

I also did some [experiment](https://github.com/pytorch/xla/issues/4686). For `I don't think we can always reshape [<=6,1] tensor to a [2, <=3]`, see case 6. It would succeed without the check https://github.com/pytorch/xla/blob/58894f0e0cc89c2a6d899a22e5619cac00cc8914/torch_xla/csrc/helpers.cpp#L272-L275. Wrt `what if it is a [3,1], there is no way for us to reshape it [2, x]`, see case 7. So I could be wrong but it looks to me the intention of the original author is: if we try to reshape/view a [<=6, 1] tensor to a shape [2, 3], the expected resulting shape would be a dynamic one [2, <=3]. But it makes lots of thing more complicated. So I agree with you that we can make it a error case when the input is dynamic and input_shape is static.